### PR TITLE
When checking for "name reuse", consider SUPERSEDED deploys as valid.

### PR DESCRIPTION
**Pull request by atombender**
Relates to #3134, but does not fix the underlying cause of releases ended up stuck in SUPERSEDED.

Also cleans up the boolean logic and fixes an ignored error returned from `env.Releases.History()`.

### CODE REVIEW

**Comment by technosophos** [comment]
I _think_ this is okay, from our backward compatibility guidelines. The SUPERSEDED state as history[0] is not a standard state. It is a bug for a release to be left in SUPERSEDED, so it is probably fine to ignore it from a compatibility standing.

However, the problem with the SUPERSEDED state is that it doesn't say anything about the state of the cluster, and one thing that the unique naming algorithm is supposed to prevent is one distinct release from "adopting" or colliding with existing resources. I guess I would feel better about this issue if I knew what situation we were in where (a) release history is in SUPERSEDED, and (b) there were no existing resources that could become naming conflicts.

Because of that, I'm hesitant to merge this PR. But some clarification on how we get into this state might help.

**Comment by atombender** [comment]
True. As I mention, this does not fix the underlying cause (see #3134, which should be reopened). Mind you, this change is needed in order to recover from *existing* releases stuck in `SUPERSEDED` that cannot be deployed because of the name reuse error.

**Comment by technosophos** [comment]
Re-reading my last message, I wasn't very clear.

What I am worried about is that allowing re-using SUPERSEDED names is that we don't know whether releases in the SUPERSEDED state still have things deployed. By allowing re-use, are we opening up the possibility for more runtime issues (which are harder to detect and fix)? Or are we better off forcing a `helm delete --purge` in these cases in order to ensure that the release is cleaned up?

Otherwise I am fine with this PR.

**Comment by atombender** [comment]
I understood that. I don't understand what the danger is, though. Doesn't Helm simply converge idempotently — create/update new resources, then delete superfluous resources — regardless?

My experience with #3134 (which has been happening almost daily; again, can you please reopen it?) is that "fixing" such a broken `SUPERSEDED` release is not destructive.

Forcing a `helm delete --purge` is a problem, since it will actually delete resources, so it is not a solution for production systems. Even if you could afford the downtime, that's also a problem for stateful resources; it will cause PVCs to be unbound, and last I checked, Helm did not have `--cascade=false`, so statefulsets will be destroyed. I really wish that you could do `helm delete --purge --preserve` that only deleted the release state, and that a subsequent `helm install --upgrade` could adopt the existing resources.

File: pkg/tiller/release_server.go [hunk 1]
```diff
@@ -168,32 +168,35 @@ func (s *ReleaseServer) reuseValues(req *services.UpdateReleaseRequest, current
...
 	if start != "" {
-
 		if len(start) > releaseNameMaxLen {
 			return "", fmt.Errorf("release name %q exceeds max length of %d", start, releaseNameMaxLen)
 		}
 
 		h, err := s.env.Releases.History(start)
 		if err != nil || len(h) < 1 {
-			return start, nil
+			return start, err
```
Comment (technosophos) [inline pkg/tiller/release_server.go]: Why change this?

Comment (technosophos) [reply pkg/tiller/release_server.go]: With this change `helm install -n foooooo $ANYTHING` fails

**Review by technosophos — CHANGES_REQUESTED** [review]
Can you back out the change that returns an error when a name is unique?

It looks like some additional error handling could be done on that call to `History()`, but you don't need to do that. Probably we should check that History() returns ErrNotFound, and swallow the error there. Otherwise we probably should return that error.

**Review by technosophos — APPROVED** [review]

**Review by technosophos — CHANGES_REQUESTED** [review]
Ugh... sorry, I approved it too early, and Circle caught the error. It looks like for some reason those errors got created as functions that generate strings, rather than as types. So you'll probably have to call the function and do an equality match on it.

**Comment by atombender** [comment]
Oops, sorry! I've force-pushed a fix that I believe is correct, although the string-based error checking is iffy. Why not use structural errors here?

My ability to do this properly was hampered a bit by a perfect storm of workspace problems — my VSCode's Go extension was misbehaving, and `glide install` wasn't working due to [this pesky Glide bug](https://github.com/Masterminds/glide/issues/210) (requiring a manual edit to the lockfile), and Go linting was taking several minutes before showing errors, presumably due to the importing of the rather huge Kubernetes packages. Things are better now, though.

**Comment by bacongobbler** [comment]
just checking in on this PR. Where are we at on this feature? Are you able to spend some time rebasing and whipping this into shape some time this week? Thanks!

**Pull request closed without merging**